_pages/dev/state_management/state-persistence.md

```diff
@@ -0,0 +1,143 @@
+---
+title: State persistence
+---
+
+In version 2.0 we introduced new mechanism to persist state of the application. Previously we only had very simple and declarative mechanism called storageSync where you could provide store properties keys to persist and those persisted keys were used on application start to set initial state in the store.
+
+## API
+
+Heart of the new feature is the `StatePersistenceService` and specifically its `syncWithStorage` method. First let's go through all the options that you can pass to this function:
+
+- `key` is used to distinguish one feature state persistence from another one. Example: to store active cart id I would use `cart` key and for user session data `session`.
+- `state$` observable which should emit every time you want to save new value to persistent storage. Example: to persist active cart id every time active cart id changes this observable should emit with new value.
+- `context$` observable describes valid context for particular state. Example: active cart id is valid only on the one base site. On different base sites we want to use different carts. So in that case for `context$` we would use observable emitting base site every time it changes.
+- `storageType` specifies storage type that should be used. By default it is local storage, but you can change that for example to session storage.
+- `onRead` callback is invoked every time context changes. In case of the cart every time we change base site this callback would be called with value read from storage for that particular context. It might dispatch `undefined` value when there wasn't anything saved in storage.
+
+## Flow of the state synchronization
+
+1. I setup my own state persistence service with:
+
+- `key`: `"cart"`
+- `state$` : pointing to active cart id selector
+- `context$`: `this.siteContextParamsService.getValues([BASE_SITE_CONTEXT_ID])` - so cart is valid only on the same base site
+- `onRead`: function dispatching `ClearCart` action and then `SetActiveCartId` when id was read from storage
+
+1. I enter `electronics` base site for the first time
+1. `onRead` was dispatched with `undefined`, because we didn't have anything under `spartacus⚿electronics⚿cart` key in local storage. Because of that we cleared cart state.
```

Comment:
```suggestion
1. `onRead` callback was invoked with `undefined`, because we didn't have anything under `spartacus⚿electronics⚿cart` key in local storage. In our implementation of `onRead` we only clear the cart state, by dispatching `ClearCart` action.
```